shared/domains/chelonia/persistent-actions.js

```diff
@@ -0,0 +1,219 @@
+  status: PersistentActionStatus
+  timer: TimeoutID | void
+
+  constructor (invocation: SbpInvocation, options: PersistentActionOptions = {}) {
+    // $FlowFixMe: Cannot resolve name `crypto`.
+    this.id = crypto.randomUUID()
+    this.invocation = invocation
+    this.options = { ...defaultOptions, ...options }
+    this.status = {
+      attempting: false,
+      failedAttemptsSoFar: 0,
+      lastError: '',
+      nextRetry: '',
+      resolved: false
+    }
+  }
+
+  async attempt (): Promise<void> {
+    // Bail out if the action is already attempting or resolved.
+    if (this.status.attempting || this.status.resolved) return
+    if (await this.trySBP(this.options.skipCondition)) {
+      this.cancel()
+      return
+    }
+    try {
+      this.status.attempting = true
+      const result = await sbp(...this.invocation)
+      this.status.attempting = false
+      this.handleSuccess(result)
+    } catch (error) {
+      this.status.attempting = false
+      this.handleError(error)
+    }
+  }
+
+  cancel (): void {
+    this.timer && clearTimeout(this.timer)
+    this.status.nextRetry = ''
+    this.status.resolved = true
+  }
+
+  async handleError (error: Error): Promise<void> {
+    const { id, options, status } = this
+    // Update relevant status fields before calling any optional code.
+    status.failedAttemptsSoFar++
+    status.lastError = error.message
+    const anyAttemptLeft = options.maxAttempts > status.failedAttemptsSoFar
+    if (!anyAttemptLeft) status.resolved = true
+    status.nextRetry = anyAttemptLeft && !status.resolved
+      ? new Date(Date.now() + options.retrySeconds * 1e3).toISOString()
+      : ''
+    // Perform any optional SBP invocation.
+    sbp('okTurtles.events/emit', PERSISTENT_ACTION_FAILURE, { error, id })
+    await this.trySBP(options.errorInvocation)
+    if (!anyAttemptLeft) {
+      sbp('okTurtles.events/emit', PERSISTENT_ACTION_TOTAL_FAILURE, { error, id })
+      await this.trySBP(options.totalFailureInvocation)
+    }
+    // Schedule a retry if appropriate.
+    if (status.nextRetry) {
+      // Note: there should be no older active timeout to clear.
+      this.timer = setTimeout(() => this.attempt(), this.options.retrySeconds * 1e3)
+    }
+  }
+
+  handleSuccess (result: any): void {
+    const { id, status } = this
+    status.lastError = ''
+    status.nextRetry = ''
+    status.resolved = true
+    sbp('okTurtles.events/emit', PERSISTENT_ACTION_SUCCESS, { id, result })
+  }
+
+  async trySBP (invocation: SbpInvocation | void): Promise<any> {
+    try {
+      return invocation ? await sbp(...invocation) : undefined
+    } catch (error) {
+      console.error(tag, error.message)
+    }
+  }
+}
+
+// SBP API
+
+sbp('sbp/selectors/register', {
+  'chelonia.persistentActions/_init' (): void {
+    this.actionsByID = Object.create(null)
+    // Necessary for now as _init cannot be async. Becomes true when 'configure' has been called.
+    this.ready = false
```

Comment:
I can't find the place to set `this.ready` `true` in `configure` function. Did you miss it by mistake?

Reply:
Indeed, good catch. BTW I've since replaced it by a direct check for availability of `databaseKey`